### DISCUSSION TITLE
Set environmental variables via npm script and command line

### DISCUSSION THREAD

**Discussion by TossExecutor** [cypress-io/cypress#14292] [CI setup] [answered]
Hello everyone,

I'm trying to follow through this guide: https://glebbahmutov.com/blog/cypress-tips-and-tricks/#pass-the-environment-variables-correctly

So I want to open cypress and also pass env variables for my tests, this should be done by using a script in my package.json.
Unfortunately, it wont work on my operating system: Instead of passing the variable's value, I get "$DB_USERNAME" and "$PASSWORD". I want to use Powershell ISE on my Windows system. I have tried many things, like using %DB_USERNAME%, but it just doesn't work for me...

Here ist my current approach:

my script lines (package.json)
```
"scripts": {
    "cy:run": "cypress run --env db.user=$DB_USERNAME,db.password=$PASSWORD",
    "cy:open": "cypress open --env db.user=$DB_USERNAME,db.password=$PASSWORD"
```

my powershell file (start.ps1):
```
npm run
$DB_USERNAME="testUser" 
$PASSWORD="123"
$DB_USERNAME, $PASSWORD | npm run cy:open
```

this my output in cypress browser (configuration tab)
```
env:
db.user: "$DB_USERNAME"
db.password: "$PASSWORD"
```

Do you have any suggestions what I am currently doing wrong here? I kinda ran out of ideas :)

**Comment by TossExecutor** [comment]
Hi, can someone help, please? :)

**Comment by tit** [comment]
@TossExecutor, try next construction:

```
$env:db.user = "testUser"; $env:db.password = "123"; cypress run --env db.user=$env:db.user,db.password=$env:db.password
```

**Reply by TossExecutor** [reply]
thanks for your reply, that doesn't work, unfortunately.

I'm assuming you meant this in my case:

package.json
```
"scripts": {
    "cy:open": "cypress open --env db.user=$env:db.user,db.password=$env:db.password
```

powershell file (.ps1)
```
$env:db.user = "testUser"; $env:db.password = "123" | npm run cy:open
```

If I run this my Console says the following for the passing values
```
The "user" property was not found for this object. Make sure that the property exists and can be set.
In line:1 Character:1
+ $env:db.user = "testUser"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyNotFound
```

**Comment by BudaPavel** [comment]
Setting --env seems not to be possible for nested keys in cypress in accordance to [documentation](https://docs.cypress.io/guides/guides/command-line.html#cypress-open).

You can try to use next construction. It's single line, but I think you would be able to handle how it works.

```bash
export DB_CONFIG='{"database":"database_name"}' && cypress open --env db=${DB_CONFIG}
```

1. We export into DB_CONFIG variable json object
2. We pass this object into cypress open

**Reply by BudaPavel** [reply]
So normaly in you package.json you need to add:
```js
"cy:open": "cypress open --env db=${DB_CONFIG}"
```

And in you power shell script you run it as:
```bash
export DB_CONFIG='{"database": "database_name", "user": "user", "password": "password"}' && npm run cy:open
```

Not tested on powershell, but works in (wsl)bash :)

**Reply by TossExecutor** [reply]
Thank you very much for your answer! Unfortunately, it's still not working for me.
I can't use `export`,  neither nor `&&` in Powershell.

I tried to work around it ended up with the following: 

```
For package.json I have the same code as you have.
```

For Powershell this:

```
$credential = $host.ui.PromptForCredential("Need credentials", "Please enter your user name and password.", "", "")
$PASSWORD = $credential.GetNetworkCredential().password
$USERNAME = $credential.GetNetworkCredential().username
Set-Variable -Name "$DB_CONFIG" -Value '{"database": "database_name", "user":"',$USERNAME,', "password": ',$PASSWORD,'}'
npm run cy:open
```

It does not throw any errors, but I still have `db: "${DB_CONFIG}"` this in my cypress configuration :(
I dunno what I do wrong here and I dunno for what I can search for addressing this problem.

**Reply by BudaPavel** [reply]
Hmm from first attempt this construction seems to work, but sets value as string and not object. But maybe would be some food to play with :)
```powershell
$env:DB_CONFIG='{"database": "database_name", "user": "user", "password": "password"}' ; cypress open --env db=$env:DB_CONFIG
```
![image](https://user-images.githubusercontent.com/69157949/104748147-6ce09080-5762-11eb-97df-c6a9e0c93596.png)

**Comment by TossExecutor** [comment]
Unfortunately, those things won't work on a windows pc and Powershell. Could it be that Powershell is not supported in a way for cypress?

**Comment by bahmutov** [comment]
It seems to me the problem is in setting the environment variables before running Cypress. Cypress just reads the variables, it has no control over setting them.

**Reply by TossExecutor** [reply]
Thank you very much for replying. To make it a bit more clear I have made a picture to explain my issue:

![2021_01_21_10_13_44_Window](https://user-images.githubusercontent.com/76519050/105331828-311b5000-5bd4-11eb-9b85-bddf22b1ad36.png)

I would like to have the first attempt working. As you can see setting the variables has worked, so I am not sure what the problem is. Do you have an idea?

**Reply by BudaPavel** [reply]
Not sure, but have you tried in cypress.json setting script as?
```js
"cy:open": "cypress open --env user=$env:$DB_USERNAME,password=$env:$DB_PASSWORD"
```

**Reply by TossExecutor** [reply]
Thanks for the patience :)
Yes, I tried it, unfortunately, get a similar result here:
![image](https://user-images.githubusercontent.com/76519050/105961439-cad47880-607e-11eb-99c4-91b30808baa8.png)

**Comment by bahmutov** [comment, marked as answer]
Seems your shell does not interpolate variables inside `npm run` scripts. I have no idea what it does or why it does not work as other shells. I have recorded a video showing my experience with passing the environment variables: https://youtu.be/jAiYePsxPl4 At the end I suggest using the plugin file to grab the environment variables and avoid relying on the shell.

**Comment by karen-campbell** [comment]
Hi @bahmutov I have watched your video and it was very informative however I am still having issues.

This is my `cypress.json` file and I am looking to set the values inside of `db` within the command line. They just seem to get set as a string no matter what I try. Do you have any advice?
```
{
  "baseUrl": "https://localhost:12345/"
  "env": 
  {
    "db":
    {
      "userName": "",
      "password": "",
      "server": "",
      "options":
      {
        "database": ""
      }
    }
  }
}
```

**Comment by bahmutov** [comment]
I have opened an issue https://github.com/cypress-io/cypress/issues/15323 about parsing JSON objects in `--env` CLI option, but for now the workaround: use a separate `CYPRESS_` variable instead of `--env` option. For example

```
$ CYPRESS_person='{ "name": { "first": "Joe", "last": "Smith" }, "age": 18 }' npx cypress open
```
